# Errores y Precisión numérica

## Definiciones y estándares con respecto a los errores y precisión numérica

Para medir los errores numéricos usamos la nomenclatura de los errores experimentales. Por ejemplo, usamos la notación de dar el valor resultante $\pm$ el error.


In [ ]:
from random import * #importar la librería de pseudoaleatorios
l = []
for i in range(1000):
    x = uniform(0.8, 2.8) # valores aleatorios float en el intervalo dado
    l.append(x)

n = len(l)
suma = 0
for i in range(n):
#    suma=suma + l[i]
    suma += l[i]
promedio = suma/n

error = max(l) - min(l)
print('El promedio resultante es: \n', promedio, '+-', error)

El promedio resultante es: 
 1.7969454382065317 +- 1.9996225301648465


### Error absoluto, error relativo y error porcentual

Usamos $X_{teo}$ para el valor teórico o verdadero de alguna variable y $X_{aprox}$ para el valor aproximado o numérico.

El error absoluto está definido por:

$$e_{abs}(X_{aprox})=|X_{teo}-X_{aprox}|$$

El error relativo es:
$$e_{rel}(X_{aprox})=\frac{|X_{teo}-X_{aprox}|}{X_{teo}}$$

El error porcentual es:
$$e_{por}(X_{aprox})=e_{rel}(X_{aprox})(100\%)$$

### Ejemplo:
Sea $X_{teo}=e$ y $X_{aprox}=\tfrac{19}{7}$

$$e_{abs}(X_{aprox})=|e - \frac{19}{7}| \approx 0.003996$$

El error relativo es:
$$e_{rel}(X_{aprox})=\frac{|e - \frac{19}{7}|}{e} \approx 0.00147$$

El error porcentual es:
$$e_{por}(X_{aprox})=(0.00147)(100\%)=1.47\%$$


In [ ]:
from math import *
def error_abs(real,approx):
    return abs(real-approx) #valor absoluto
def error_rel(real,approx):
    return (abs(real-approx))/real
def error_por(real,approx):
    return ((abs(real-approx))/real)*100

x_teo=e
x_aprox=19/7
#x_teo=float(input('Da el valor teorico'))
#x_aprox=float(input('Da el valor numerico'))

e_abs=error_abs(x_teo,x_aprox)
e_rel=error_rel(x_teo,x_aprox)
e_por=error_por(x_teo,x_aprox)

print('El error absoluto es : %5.4f' %e_abs)
print('El error relativo es : %5.4f' %e_rel)
print('El error porcentual es : %7.6f' %e_por)

El error absoluto es : 0.0040
El error relativo es : 0.0015
El error porcentual es : 0.147009


## Digitos significativos

### Definición:
Decimos que $X_{aprox}$ tiene $m$ digitos significativos cuando el error de $X_A$ cumple con $err(X_{aprox})\leq tol$ en el digito (m+1) distintos de cero.

### Ejemplo
$$e_{abs}(X_{aprox})=\left|e-\frac{19}{7}\right|\approx 0.00\mathbf{3}996$$
se tiene 3 digitos significativos

## Tipos de errores
1. Errores numéricos derivados de la aproximación del método numérico $tol$
2. Errores numéricos de la aritmética discreta de la computadora $X_{aprox}$.
3. Errores en el planteamiento del modelo (¡Vacas esféricas, cuidado!)

## 1. Errores numéricos de la aproximación

Aquí consideramos los términos que despreciamos de las aproximaciones numéricos.

Serie de Taylor
$$e^x \approx 1 + x + \frac{1}{2} x^2$$

Diferenciación
$$ f'(x) \approx \frac{f(x + h) - f(x - h)}{2h}$$

Integración numérica
$$ \int_0^1 f(x) dx \approx \frac{1}{n} \sum_{j=1}^n f(\frac{j}{n})$$



## 2. Errores numéricos de la aritmética discreta

Hay que considerar que la computadora tiene limitaciones. Las computadoras no pueden guardar números con decimales infinitos, por lo que no se puede manipular elementos continuos, sino que toda la matemáticas que se hagan con la computadora serán discretas. Tampoco números muy grandes o muy chicos.

In [ ]:
x=1.1+2.2
print(x)
print(x==3.30000000000000)

3.3000000000000003
False


In [ ]:
x=1.1+2.2
if x==3.3:
    print(x)
else:
    print('x distinto a 3.3')
#¿Por qué no se imprimió x terminando el "if"? Ocurre que por error de redonde
#1.1+2.2 no es exactamente 3.3, sino que es 3.300000000000003. Por lo cual no
#se cumple la condición del if. Esto es bastante común en física computacional,
#por lo que la mejor manera de tratar esto, es introducir un valor
#epsilón en el código.
print(x)

x distinto a 3.3
3.3000000000000003


In [ ]:
x=1.1+2.2
epsilon=1e-12 #toleracion
if abs(x-3.3)<epsilon:
    print(x)

3.3000000000000003


En Python, el nivel de precisión estandar es de 16 cifras significativas. Esto significa que los números irracionales serán truncados:

Valor verdadero de Pi: 3.1415956535897932384626...

Valor en Python de Pi: 3.141595653589793

Diferencias:           0.0000000000000002384626

A este error se le llama error de redondeo.

Entonces, el error numérico de un número, que se denotará $\epsilon$ (epsilon), se define como la cantidad calculada por la computadora menos el valor real (el error absoluto).

En general, si $X_{aprox}$ es una aproximación a cierto número de cifras significativas, digamos 16, entonces el error de redondeo tiene un valor típico de $\frac{X_{aprox}}{10^{16}}$. Usualmente es una buena aproximación considerar el error como una cantidad uniformemente distribuida de un número con la desviación estandar $\sigma=CX_{aprox}$, donde $C\approx 10^{-16}$. Nos referimos a $C$ como una constante de error.

Comúnmente nos referimos a $\epsilon$ como la desviación estandar ya que sí dieramos un valor exacto de $\epsilon$ sería muy fácil obtener el valor real de x.

Veamos con más cuidado esto de los errores y la precisión de los cálculos:

Sean:
\begin{eqnarray*}
        x&=&1000000000000000.0000000000000\\
        y&=&1000000000000001.234567890123478
\end{eqnarray*}
Recordando que la computadora sólo guarda 16 cifras significativas, y será truncada. La diferencia entre y-x=1.2, cuando en realidad es 1.234567890123478, lo cual el error porcentual es del 0.23%. Sin embargo, al agregar una cifra decimal a x y y, el error porcentual se incrementa considerablemente.

Veamos en código como se comporta todo.

In [ ]:
x = 1000000000000000
y = 1000000000000001.234567890123478
a = y - x
error_real = 1.234567890123478
print(a)
#error = (abs(a)/(error_real))*100
error = (abs(a-error_real)/(error_real))*100
print(error, '%')

1.25
1.2500009112482682 %


Una manera de aumentar las cifras decimales asignadas a un flotante y disminuir el error por redondeo es usar el módulo decimal.

In [ ]:
#Primero, veremos el paquete decimal
from decimal import getcontext,Decimal
from math import sqrt,exp,pi

print(1/7)
print(pi)
getcontext().prec = 6
print(Decimal(1) / Decimal(7))
getcontext().prec = 60
print(1/7)
print(Decimal(pi))
print(Decimal(1) / Decimal(7))
print(Decimal(1/7))
print(getcontext().sqrt(Decimal(2)))
#print(getcontext().sqrt(Decimal(2)))

0.14285714285714285
3.141592653589793
0.142857
0.14285714285714285
3.141592653589793115997963468544185161590576171875
0.142857142857142857142857142857142857142857142857142857142857
0.142857142857142849212692681248881854116916656494140625
1.41421356237309504880168872420969807856967187537694807317668


### Ejemplo

Evaluamos
$$f(x) = x(\sqrt{x + 1} - \sqrt{x})$$
con una calculadora de 6 digitos de precisión.

In [ ]:
from decimal import Decimal as D, getcontext as C #C=getcontext()
import pandas as pd
from math import sqrt

C().prec = 6 #get.context().prec=6

def f(x):#getcontext es 16 cifras significativas
    return x * (sqrt(x + 1) - sqrt(x))
def fd(x): #getcontext es de 6 cifras significativas
    return x * (C().sqrt(x + D(1)) - C().sqrt(x))
residuo = []
idx = []
for i in range(10):
    x = 10**i
    idx.append(x)
    residuo.append([fd(D(x)), f(x)])
#print(residuo)
cols = ["Aprox", "Real"]
pd.DataFrame(residuo, index=idx, columns=cols)

Aprox          Real
1           0.41421      0.414214
10          1.54340      1.543471
100          4.9900      4.987562
1000        15.8000     15.807437
10000        50.000     49.998750
100000      100.000    158.113488
1000000           0    499.999875
10000000       0.00   1581.138790
100000000         0   5000.000056
1000000000      0.0  15811.390767

### Ejemplo

Definimos
$$f_1(x) = \frac{1 - \cos(x)}{x^2}, \qquad x \not= 0$$
evaluamos con 10 digitos de precisión

In [ ]:
from decimal import Decimal as D, getcontext as C #C=getcontext()
import pandas as pd
from math import sqrt, cos

C().prec = 10
#from math import cos
def f1(x):
    return (1 - cos(x)) / (x*x)
def f1d(x):
    #return D(D(1) - D(cos(D(x)))) / (D(x*x)) #regresa en el contexto default 16
    return +D(D(1) - (+D(cos(+D(x))))) / (+D(x)*+D(x))
##feb 2024 del modulo (3.12.2) + -D indicador de la prec sustituye el getcontext
residuo = []
idx = []
for i in range(1, 7):
    x = 10**(-i)
    idx.append(x)
    err=abs(float(f1d(+D(x)))-f1(x))
    residuo.append([f1d(+D(x)), f1(x),err*100])
#pd.set_option("precision", C().prec)
cols = ["Aprox", "Real","Error%"]
pd.DataFrame(residuo, index=idx, columns=cols)
#pd.DataFrame(residuo, columns=cols)

Aprox      Real        Error%
0.100000  0.49958347  0.499583  2.197418e-07
0.010000    0.499996  0.499996  1.666526e-05
0.001000         0.5  0.500000  4.167450e-06
0.000100         0.5  0.500000  3.038735e-07
0.000010           1  0.500000  5.000000e+01
0.000001       0E+12  0.500044  5.000445e+01

In [ ]:
x = +D(0.001)
+D(cos(x))  #Redondeamos el resultado final regresando a la precisión default

Decimal('0.9999995000')

### 3. Errores en el planteamiento

Es consecuencia de restar dos numeros casí iguales, generan perdidas de significancia. Es difícil de detectar y no siempre es fácil evitarlo.

###  Ejemplo:
Sea $f(x) = \frac{1 - \cos(x)}{x^2}$, usando $$\cos(2\theta) = 2\cos^2(\theta) - 1 = 1 - 2\sin^2(\theta)$$

Con $x=2\theta$:
\begin{eqnarray}
f(x) &=& \frac{1 - \cos(x)}{x^2} = \frac{2\sin^2(x/2)}{x^2}\\
     &=& \frac{1}{2} \left[\frac{\sin(x/2)}{x/2}\right]^2
\end{eqnarray}

Comparar varios valores muy pequeños $x\approx 0.001$

In [ ]:
from decimal import Decimal as D, getcontext as C
from math import sin,cos

C().prec = 10

def f1(x):
    return (1 - cos(x)) / (x*x)
def f1d(x):
    return +D(D(1) - (+D(cos(+D(x))))) / (+D(+D(x)*+D(x)))

def f1a(x):
    return 0.5 * (sin(x/2) / (x/2))**2
def f1da(x):
    return +D(0.5) * (+D(sin(x/2))/ (x/2))**2

print('Funcion original y 16 digitos',+D(f1(0.00001)))
print('Funcion aprox y 16 digitos',+D(f1a(0.00001)))
print('Funcion original y 10 digitos',f1d(+D(0.00001)))
print('Funcion aprox y 10 digitos',f1da(+D(0.00001)))

#f1 sin aproximar y con 16 digitos
#f1a sin aproximar y con 16 digitos
#f1d con aproximacion y con 10 digitos
#f1da con aproximacionr y con 10 digitos

Funcion original y 16 digitos 0.5000000414
Funcion aprox y 16 digitos 0.5000000000
Funcion original y 10 digitos 1
Funcion aprox y 10 digitos 0.500


Notamos que el resultado approximado y el correcto están mucho más cerca, el error casi desapareció.

In [ ]:
res = []
idx = []
for i in range(1, 7):
    x = 10**(-i)
    idx.append(x)
    res.append([f1d(+D(x)), f1da(+D(x)), f1(x), f1a(x)])
cols = ["Apr cos", "Apr sen", "Exacto cos", "Exacto sen"]
#pd.set_option("precision", C().prec)
pd.DataFrame(res, index=idx, columns=cols)


Apr cos       Apr sen  Exacto cos  Exacto sen
0.100000  0.49958347  0.4995834722    0.499583    0.499583
0.010000    0.499996  0.4999958334    0.499996    0.499996
0.001000         0.5  0.4999999584    0.500000    0.500000
0.000100         0.5  0.4999999996    0.500000    0.500000
0.000010           1         0.500    0.500000    0.500000
0.000001       0E+12         0.500    0.500044    0.500000

### Ejemplo (más sutil)

Evaluamos $e^{-5}$ usando la aproximación de Taylor
$$e^{-5} \approx 1 + \frac{(-5)}{1!} + \frac{(-5)^2}{2!} + \frac{(-5)^3}{3!} + \dots + \frac{(-5)^n}{n!}$$

Usando $n = 25$, el error es de
$$\left| \frac{(-5)^{26}}{26!} e^c \right| \le 10^{-8}$$
Lo valuamos con 4 digitos de precisión.

In [ ]:
from math import exp, factorial
C().prec = 4
res = +D(1)
for i in range(1, 26):
    res = res + (+D((-5)**i/factorial(i)))

(res, +D(exp(-5)), +D(exp(-5)) - res)

(Decimal('0.009985'), Decimal('0.006738'), Decimal('-0.003247'))

Notamos la diferencia grande entre el resultado correcto 0.006738 y la aproximación 0.009985, con un error de -0.003247.

Para entender el efecto analizamos el termino
$$\frac{(-5)^3}{3!} = - \frac{125}{6} \approx -20.83$$

In [ ]:
term = (-5)**3 / factorial(3)
aprox = +D(term)
error = +D(term - float(aprox))
(term, aprox, error)

(-20.833333333333332, Decimal('-20.83'), Decimal('-0.003333'))

Veamos que el error de -0.003333 es del mismo orden de magnitud que el resultado fínal de 0.006738.

**Sumar (restar) terminos con un resultado mucho menor que los terminos individuales resulta en una perdida importante de significancia.**

## Underflow y Overflow
Python permite tener números lo suficientemente grandes o chicos, alrededor de 16 cifras significativas. Por ejemplo, el valor flotante más largo que se puede ingresar es $10^{308}$. Este número suele ser suficiente para hacer física computacional, pero en ocasiones, no es suficiente. Si nos pasamos de este número, ocurre que la variable se sobresatura "overflowed" y Python nos regresará un valor "inf", llevando a errores numéricos graves o a que el código no pueda ser ejecutado.

(Nota: Para números muy pequeños, como $2^{-1022}$)

In [ ]:
def factorial_entera(n):
    multiplicacion=1 #Funcion entera
    for i in range(1,n):
        multiplicacion*=i
    return multiplicacion
def factorial_flotante(n):
    multiplicacion=1.0 #Funcion flotante
    for i in range(1,n):
        multiplicacion*=i
    return multiplicacion

print('*El factorial entero es :\n',factorial_entera(1000)) #200000
print('**El factorial flotante es :\n',factorial_flotante(174)) #overflow #171 límite Colab

*El factorial entero es :
 4023872600770937735437024339230039857193748642107146325437999104299385123986290205920442084869694048004799886101971960586316668729948085589013238296699445909974245040870737599188236277271887325197795059509952761208749754624970436014182780946464962910563938874378864873371191810458257836478499770124766328898359557354325131853239584630755574091142624174743493475534286465766116677973966688202912073791438537195882498081268678383745597317461360853795345242215865932019280908782973084313928444032812315586110369768013573042161687476096758713483120254785893207671691324484262361314125087802080002616831510273418279777047846358681701643650241536913982812648102130927612448963599287051149649754199093422215668325720808213331861168115536158365469840467089756029009505376164758477284218896796462449451607653534081989013854424879849599533191017233555566021394503997362807501378376153071277619268490343526252000158885351473316117021039681759215109077880193931781141945452572238655414

### Ejercicio:
Considera dos números
$$x=1$$
$$y=1+1\times10^{-14}\sqrt(2)$$

Haz un programita en Python que calcule directamente
$$\sqrt(2)$$
y
$$\sqrt(2)=1\times10^{14}(y-x).$$
¿Qué valor te da? ¿Cuál es el valor porcentual?

In [ ]:
#Ejercicio


1.4210854715202004
1.4142135623730951
